CNN Triplet Tranining

Wangyida opened the pull request:
* 3D object recognition and face recognition samples
* 2 types of triplet loss for loss
* free number of sets for triplet loss

Wangyida commented:
@bhack @shelhamer @mtamburrano This is a newly rebased PR for triplet loss, I just added a face recognition sample in the examples for general using.
This is related to PR #2603

shelhamer commented:
@Wangyida please do not keep opening new PRs instead of updating your existing PR. You can update a PR at any time by force pushing `git push -f` to your fork's branch. This makes it easier to track what is going on.

bhack commented:
And I think you are still incorporating changes that you don't want to propose hear. Review file changes of this PR. Do you really want this?

bhack commented:
Probably you want to razionalize the two commits.

Wangyida commented:
Now two more changes in file mode is restored, I accidentally reset a commit version earlier than my first commit in the other PR, so I reopened here, it might also be able to reopen the disabled PR?

bhack commented:
Ok  changes seems isolated now. I think you can continue here. Next time close and reopen a new one only in particular cases.

bhack commented:
Is this ready for a review pass now?

Wangyida commented:
@bhack  Yes, sure~

bhack commented:
@shelhamer Can you make a first quick pass?

IMG-PRCSNG commented:
On using your triplet loss layer, I encountered an error which raised a SIGSEGV
src/caffe/layers/triplet_loss_layer.cu - Line number 36
` loss += dist_sq_pos.gpu_data()[i];`

The problem seems to be arising on using `gpu_data`. Was wondering if you have faced this problem before.
I will poke around a bit more and get back to you.

Wangyida commented:
@IMG-PRCSNG I have run the GPU codes on K40 and there seems no problem in the latest codes, I used to encounter some problems in the gpu_data of some metric operation functions, so I changed it to a cpu_data as the same data are stored also which could be used.

IMG-PRCSNG commented:
Currently I am using only the CPU part of the code which doesn't throw any error. I find no reason why the GPU code should fail. I will try it out with cpu_data as you suggested. Thanks!

swamiviv commented:
@Wangyida  Great job on the triplet loss. I am slightly confused though about how the filelists for the Image Data layer should be created. Can you provide an example ? Are the triplets generated online per batch ?

swamiviv commented:
@Wangyida Great job on the code! 
Are the inputs to the Imagedata layer just regular filelists ? How is the triplet generation done ?

Wangyida commented:
@swamiviv  As for the object pose related issue, you can have a look in the code in the training data generation code using leveldb, and for other free style using manner of training data, you can just have a reference on this one, it is the same algorithm on training samples arrangement style using filelist #2982

swamiviv commented:
Thanks a lot!

On Thu, Oct 15, 2015 at 10:43 PM, Yida Wang <notifications@github.com>
wrote:

> @swamiviv <https://github.com/swamiviv> As for the object pose related
> issue, you can have a look in the code in the training data generation code
> using leveldb, and for other free style using manner of training data, you
> can just have a reference on this one, it is the same algorithm on training
> samples arrangement style using filelist #2982
> <https://github.com/BVLC/caffe/issues/2982>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BVLC/caffe/pull/3123#issuecomment-148582788>.
>



-- 
--Swami

cysin commented:
HI Yida, I tried your triplet loss implementation but got errors when using GPU mode:
```
I1026 10:25:52.054105  3953 blocking_queue.cpp:50] Data layer prefetch queue empty
*** Aborted at 1445826352 (unix time) try "date -d @1445826352" if you are using GNU date ***
PC: @     0x7f1b9ca65237 caffe::TripletLossLayer<>::Forward_gpu()
*** SIGSEGV (@0x7011c5000) received by PID 3953 (TID 0x7f1b9cfe4a40) from PID 18632704; stack trace: ***
    @     0x7f1b9c1c5130 (unknown)
    @     0x7f1b9ca65237 caffe::TripletLossLayer<>::Forward_gpu()
    @     0x7f1b9c9db501 caffe::Net<>::ForwardFromTo()
    @     0x7f1b9c9db887 caffe::Net<>::ForwardPrefilled()
    @     0x7f1b9ca04fff caffe::Solver<>::Test()
    @     0x7f1b9ca058ae caffe::Solver<>::TestAll()
    @     0x7f1b9ca059e9 caffe::Solver<>::Step()
    @     0x7f1b9ca06335 caffe::Solver<>::Solve()
    @           0x40bbbb train()
    @           0x409481 main
    @     0x7f1b8fd4daf5 __libc_start_main
    @           0x409c1d (unknown)
Segmentation fault (core dumped)
```

swamiviv commented:
I get the same error too. Was it resolved ?

On Sun, Oct 25, 2015 at 10:26 PM, cysin <notifications@github.com> wrote:

> HI Yida, I tried your triplet loss implementation but got errors when
> using GPU mode:
>
> I1026 10:25:52.054105  3953 blocking_queue.cpp:50] Data layer prefetch queue empty
> *** Aborted at 1445826352 (unix time) try "date -d @1445826352" if you are using GNU date ***
> PC: @     0x7f1b9ca65237 caffe::TripletLossLayer<>::Forward_gpu()
> *** SIGSEGV (@0x7011c5000) received by PID 3953 (TID 0x7f1b9cfe4a40) from PID 18632704; stack trace: ***
>     @     0x7f1b9c1c5130 (unknown)
>     @     0x7f1b9ca65237 caffe::TripletLossLayer<>::Forward_gpu()
>     @     0x7f1b9c9db501 caffe::Net<>::ForwardFromTo()
>     @     0x7f1b9c9db887 caffe::Net<>::ForwardPrefilled()
>     @     0x7f1b9ca04fff caffe::Solver<>::Test()
>     @     0x7f1b9ca058ae caffe::Solver<>::TestAll()
>     @     0x7f1b9ca059e9 caffe::Solver<>::Step()
>     @     0x7f1b9ca06335 caffe::Solver<>::Solve()
>     @           0x40bbbb train()
>     @           0x409481 main
>     @     0x7f1b8fd4daf5 __libc_start_main
>     @           0x409c1d (unknown)
> Segmentation fault (core dumped)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BVLC/caffe/pull/3123#issuecomment-151004554>.
>



-- 
--Swami

Wangyida commented:
@cysin OK, Let me have a check on the GPU codes.

bhack commented:
/cc @vpisarev for https://github.com/Itseez/opencv_contrib/pull/276

mtamburrano commented:
I see that you try to access on `cpu_data` when you make the dot operation to calculate the loss.
I think that all the pointers should refer to gpu, I don't think that the underlying cblas function can handle mixed cpu and gpu pointers in that way

Wangyida commented:
@mtamburrano I encountered some GPU memory issues that time when I using the matrix dot function defined in CUDA, so I change to use cpu_data instead cause this function seems able to have a cpu_data backup at the same time when it is used.

bhack commented:
I see that caffe_gpu_dot call cublasSdot or cublasDdot. How that could handle host pointers?

Wangyida commented:
@bhack  OK, I will have a deeper check on it, I used to succeeded on a K80 server with this code and I find that caffe_gpu_dot is seldom used in all GPU codes of layers.

bhack commented:
@Wangyida Yes it is used but I cannot find any layer that use it with CPU pointers cause it is a wrapper call on cublas.

ljw3351639 commented:
HI Yida, I also get the error in GPU mode too. Could you help check the gpu code?

*** Aborted at 1447318398 (unix time) try "date -d @1447318398" if you are using GNU date ***
PC: @     0x7f45d8234110 caffe::TripletLossLayer<>::Forward_gpu()
*** SIGSEGV (@0x706d4ae00) received by PID 321 (TID 0x7f45d8a2a800) from PID 114601472; stack trace: ***
    @     0x7f45d6fc3d40 (unknown)
    @     0x7f45d8234110 caffe::TripletLossLayer<>::Forward_gpu()
    @     0x7f45d80b31b2 caffe::Net<>::ForwardFromTo()
    @     0x7f45d80b33df caffe::Net<>::ForwardPrefilled()
    @     0x7f45d8130ebd caffe::Solver<>::Test()
    @     0x7f45d813194e caffe::Solver<>::TestAll()
    @     0x7f45d8131a9a caffe::Solver<>::Step()
    @     0x7f45d81327f1 caffe::Solver<>::Solve()
    @           0x40a6ce train()
    @           0x406908 main
    @     0x7f45d6faeec5 (unknown)
    @           0x406e28 (unknown)
    @                0x0 (unknown)
Segmentation fault (core dumped)

mtamburrano commented:
hi @ljw3351639, can you test if it is enough to replace all the occurrences of ` mutable_cpu_data()` with `mutable_gpu_data()` in [src/caffe/layers/triplet_loss_layer.cu](https://github.com/Wangyida/caffe/blob/cnn_triplet/src/caffe/layers/triplet_loss_layer.cu)?

ljw3351639 commented:
@Wangyida @mtamburrano I have done it, but it also have some problems.
*** Aborted at 1447333719 (unix time) try "date -d @1447333719" if you are using GNU date ***
PC: @     0x7f979c1ecaec (unknown)
*** SIGSEGV (@0x706d4b400) received by PID 19155 (TID 0x7f979fbf2800) from PID 114603008; stack trace: ***
    @     0x7f979e18bd40 (unknown)
    @     0x7f979c1ecaec (unknown)
    @     0x7f979f3dc8a7 caffe::caffe_gpu_dot<>()
    @     0x7f979f3fc100 caffe::TripletLossLayer<>::Forward_gpu()
    @     0x7f979f27b1b2 caffe::Net<>::ForwardFromTo()
    @     0x7f979f27b3df caffe::Net<>::ForwardPrefilled()
    @     0x7f979f2f8ebd caffe::Solver<>::Test()
    @     0x7f979f2f994e caffe::Solver<>::TestAll()
    @     0x7f979f2f9a9a caffe::Solver<>::Step()
    @     0x7f979f2fa7f1 caffe::Solver<>::Solve()
    @           0x40a6ce train()
    @           0x406908 main
    @     0x7f979e176ec5 (unknown)
    @           0x406e28 (unknown)
    @                0x0 (unknown)
Segmentation fault (core dumped)

Wangyida commented:
@ljw3351639 As Mannuele suggested, this codes has been modified from gpu_data to cpu_data by myself due to memory issues. Problem exists in This dot function I will have a replace on it this Sunday, I am attending CCBR2015 in Tianjin now

swamiviv commented:
@Wangyida  Did you get a chance to resolve the code ?

On Fri, Nov 13, 2015 at 1:03 AM, Yida Wang <notifications@github.com> wrote:

> @ljw3351639 <https://github.com/ljw3351639> As Mannuele suggested, this
> codes has been modified from gpu_data to cpu_data by myself due to memory
> issues. Problem exists in This dot function I will have a replace on it
> this Sunday, I am attending CCBR2015 in Tianjin now
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BVLC/caffe/pull/3123#issuecomment-156332496>.
>



-- 
--Swami

Strateus commented:
@Wangyida, can FaceNet (from Google CVPR 2015 paper: http://www.cv-foundation.org/openaccess/content_cvpr_2015/app/1A_089.pdf) be trained using this fork?

Wangyida commented:
@ljw3351639 @mtamburrano @bhack GPU codes for triplet loss is fixed and tested, it has been just pushed.

Wangyida commented:
@Strateus triplet loss could surely do this, but there might be a problem: a single pair wise term is joint together with several triplet sets, so your reference sample and positive samples should not differs too much. You could just remove the pair wise term in the triplet loss if the loss don't decrease as expected.

ljw3351639 commented:
@Wangyida  Thanks very much. It is succeed when i set losstype = 1and GPU mode. But it also have problem when i set losstype = 0 and GPU mode.

Wangyida commented:
@ljw3351639 It's fixed on loss type 0.

cbalint13 commented:
@Wangyida ,

Can force *sqash* all ?

Here is how to do it:

1) wget https://github.com/BVLC/caffe/pull/3123.patch (save save)
2) git clone https://github.com/Wangyida/caffe
3) cd caffe
4) git pull https://github.com/BVLC/caffe.git
5) git push
6) git checkout cnn_triplet
7) git reset --hard origin/master
8) git apply ../3123.patch --reject
9) git commit -a -m "Implement CNN Triplet training."
10) git push --force

- It will look like a single commit at the end.

Wangyida commented:
@cbalint13 @bhack I updated this code and add the a new optional pair wise term, all check passed, now all examples are based on pascal3d and multiple, the name list should be arranged and the file is not so little, user could arrange them easily.
@mtamburrano  I find that by properly use the pair wise term(I made it as a new optional parameter), accuracy of recognition on real images is much better~ Maybe some segmentation on real images are also meaningful.

jackculpepper commented:
This looks useful! I hope you'll finish it off.

Can you describe the different loss types? Maybe update the equations in `include/caffe/layers/triplet_loss_layer.hpp`?

I understand how you are arranging the anchor, positive, and negative samples in the batch, but it took me a while.. It would be nice to have a word or two about that in the readme.

The travis build is failing because of lint -- there are some minor tab/space problems.

The file `src/caffe/test/test_triplet_loss_layer` is missing the `.cpp` extension.

Wangyida commented:
@jackculpepper My description is not enough indeed, especially you should mind that the negative samples could be any number. There are 4 type of loss function which suit for different data for training, the first two is stable which is modified from published paper and the last 2 is used in my own paper which might not be so stable for particular data.
I will maintain the codes including demos, glad that some one benefit from it:)

jackculpepper commented:
Thanks for writing back! I'm reading through your code carefully now. There seem to be 3 loss types: 0, 1, and 2. With `use_pair = 0`, type 0 is:

```
L(x_a, x_p, x_n) = max(0, m - || x_a - x_n ||_2^2 + || x_a - x_p ||_2^2) ,
```

loss type 1 is:

```
L(x_a, x_p, x_n) = max(0, 1 - || x_a - x_n ||_2^2 / (|| x_a - x_p ||_2^2 + m)) ,
```

and loss type 2 is:

```
L(x_a, x_p, x_n) = max(0, 1 - exp(|| x_a - x_n ||_2^2) / (exp(|| x_a - x_p ||_2^2) + m)) .
```

Is this right?

Types 0 and 1 are mentioned in [Wohlhart and Lepetit, 2015](http://lrs.icg.tugraz.at/pubs/wohlhart_cvpr15.pdf) -- type 1 is equation 2 from this paper, and type 0 is mentioned in the text just below equation 2. Loss type 0 is also the same loss function as is used in [Schroff et al, 2015](http://www.cv-foundation.org/openaccess/content_cvpr_2015/app/1A_089.pdf) and [Parkhi et al, 2015](https://www.robots.ox.ac.uk/~vgg/publications/2015/Parkhi15/parkhi15.pdf).

Is type 2 from your paper? Is it under review?

Since the distances are part of all three loss types, perhaps those calculations should be factored out to reduce duplication and make the differences between types more evident. For example, in [triplet_loss_layer.cpp] (https://github.com/Wangyida/caffe/blob/cnn_triplet/src/caffe/layers/triplet_loss_layer.cpp) it looks like the only difference between loss type 0 and loss type 1 is [line 84](https://github.com/Wangyida/caffe/blob/cnn_triplet/src/caffe/layers/triplet_loss_layer.cpp#L84) vs [lines 112](https://github.com/Wangyida/caffe/blob/cnn_triplet/src/caffe/layers/triplet_loss_layer.cpp#L112) and [lines 126-127](https://github.com/Wangyida/caffe/blob/cnn_triplet/src/caffe/layers/triplet_loss_layer.cpp#L126-L127).

Thanks for making your code available!

Wangyida commented:
@jackculpepper Thank for your tips for making the codes cleaner, it's indeed seldom different in codes within different loss type, maybe code reuse could make my codes much shorter. Type 2 is not a proper one indeed because it just suit for data which differs little. Type 3 is still local available which is in review of my paper, after that I will push it upstream.
Another problem is that my local codes including local connected layer which is another one's pull request which I'm afraid that make the whole code a mess, so I still haven't updated my codes till now.

muneebshahid commented:
hello everyone, thanks for the work @Wangyida. do we have an update on this pr?. 
p.s from [this thread](https://github.com/BVLC/caffe/issues/2982#issuecomment-135326657) it seems like triplet loss only relies on how you arrange the data to calculate the loss and completely ignores the labels, and I correct in thinking so?

Wangyida commented:
@muneebshahid Codes works, the triplet is only dependent on the arrangement of samples, negative samples could be more than one, new codes contain some other layer including locally conneted layer, should I have an update on this pr?

cbalint13 commented:
@Wangyida ,

* Could add  @jackculpepper 's doc references (with some cite where is possible) about the mentioned three loss functions ? Perhaps some more little docs for the code where it would be possible, e.g. explain a bit more the arrangement of batch entries within the data conversion tool.

* Could update please bot/build status ? (I see that you need only to assign triplet_loss_param=143 instead of actual 141 value that is in use).

To update, more elegant would be to use push --force method (or rebase if works smooth for you). With push --force (described step-by-step here) you loose the logs but will end up with single compact patch ready to merge.

I am really keen to see this merged, and I consider this an important loss function type within the field of deep learning, it is the perfect recepit to cook up any kind of descriptors. I believe by regular update of this PR for compile/buildbot status will maxout all the chances to see this merged. If You could also add that older mnist example (like the one in the early state of PR) plus a ipython demonstrator for mnist case (that would proof triplet loss strong manifold arrangement capability, like e.g. in case of siamese networks), it would be the cherry of the cherries :100: !

@shelhamer, We hope this PR is within the interest of caffe. Could review/update/lead a bit this PR ? I believe @Wangyida did a great job implementing triplet-training, we could help him to place final nails on it, especially with little attention from your side.

Thanks @Wangyida !

muneebshahid commented:
@Wangyida thanks a lot for the reply. I was under the impression that something critical still remains to be implemented and that's why the pr had'nt been merged yet. 
Lastly I have one more question and may be a suggestion as well, just to be clear here is what my test file for imagedatalayer and my parameters look like.

1_reference_image label
1_simillar_image label
1_negative_image_0 label
1_negaive_image_1 label

2_reference_image label
2_simillar_image label
2_negative_image_0 label
2_negaive_image_1 label

and so on, and my parameters are 
**triplet_loss_param**
{
    **margin: 1**
    **losstype: 1**
    **num_triplets: 2**
    **use_pair: 1**
  }
with **batchsize** set to **256**. (256 % (2 + num_triplets) == 0)
After having looked at your code to me these params look ok. If you think these params are indeed correct then I think It would be better to rename the param **num_triplets** to **num_negatives** may be? as it really is the number of negative images per set. 
also from the code it looks like use_pair also incurs loss between reference and similar image for each set in addition to the normal loss incurred by relative distances?

Anyways thanks a lot for all the work and sharing your code!

Wangyida commented:
@cbalint13  Thanks for advices, I am removing redundant layer used in my codes, and the photo list will be updated, as triplet arrangement is set in image data list style for convenient arrangement, the db database conversion codes are not necessary now. I will tidy it as soon as possible~
@muneebshahid You advices are necessary, thanks for informing me! The pair wise term is only used for making euclidean distance between reference samples and positive samples smaller for noised data aiming at some particular issues such as image recovery.

cbalint13 commented:
@Wangyida , all my respect goes to you !
@shelhamer , I hope this bring all the attention about disscouses  !

In the great hopes, for the greater.

Wangyida commented:
@cbalint13 I have some local modification and codes build alright locally, but some format problem occurs in buildbot, do I have some method to check these inappropriate format locally? It's a little  tedious  checking them online.

cbalint13 commented:
@Wangyida ,

 I am afraid you can check only online. By using ```push --force``` you can check many versions and finally leave the appropriate one while all rest logs+entries are deleted. I used to do such things for e.g. opencv.

Wangyida commented:
@cbalint13 OK, Thanks:)

raingo commented:
@Wangyida What's the embedding you got for mnist example? Could you please share something like https://github.com/BVLC/caffe/blob/master/examples/siamese/mnist_siamese.ipynb

shelhamer commented on src/caffe/proto/caffe.proto:
```diff
@@ -1184,6 +1193,7 @@ message V1LayerParameter {
     SPLIT = 22;
     SLICE = 33;
     TANH = 23;
+    TRIPLET_LOSS = 40;
```
In my opinion this isn't needed since the `TripletLossLayer` was not in Caffe while the V1 protobuf were in use. @jeffdonahue @longjon should we have some kind of backward compatibility in this case?

shelhamer commented on src/caffe/util/upgrade_proto.cpp:
```diff
@@ -931,6 +931,8 @@ const char* UpgradeV1LayerType(const V1LayerParameter_LayerType type) {
     return "WindowData";
   case V1LayerParameter_LayerType_THRESHOLD:
     return "Threshold";
+  case V1LayerParameter_LayerType_TRIPLET_LOSS:
```
No need for this check -- see comment on `caffe.proto`.

shelhamer commented on cmake/Modules/FindvecLib.cmake:
```diff
@@ -23,7 +23,7 @@ find_package_handle_standard_args(vecLib DEFAULT_MSG vecLib_INCLUDE_DIR)
 
 if(VECLIB_FOUND)
   if(vecLib_INCLUDE_DIR MATCHES "^/System/Library/Frameworks/vecLib.framework.*")
-    set(vecLib_LINKER_LIBS -lcblas "-framework vecLib")
```
This should be a separate PR for CMake on OS X if you could kindly split this out.

shelhamer commented:
@Wangyida @cbalint13 to check the format locally run `make lint`; see the [developing and contributing guide](http://caffe.berkeleyvision.org/development.html).

Regarding your examples, I think the MNIST triplet example suffices and is more convenient since that data is already used in other examples. How about dropping the other examples (multiple and PASCAL) or keeping these at your fork with more detail instead?

bhack commented:
This work was also a counterpart financed with OpenCV GSoC  where example are more entangled with 3d stuffs. So I hope that @Wangyida could maintain a Pascal3d model and prototxt in model Caffe model zoo forking only prototxt and not the code. /cc @vpisarev. 
@pwohlhart Do you have got some interesting results with your student on Pascal3d without diverging too much from this PR?

Wangyida commented:
@raingo Hi, I removed an training example for MNIST because the arrangement style changed since the number of negative samples is an optional parameter. But the triplet convert tool in examples/triplet which is designed for converting binary files into leveldb files could be modified for MNIST database, I wrote it for the generated data from my OpenCV PR which is already merged in opencv_contrib distribution.
@shelhamer Thanks for your reviews in detail about the caffe.photo, I removed it. As for the training method for MNIST, I think it's not so convenient converting the binary file into leveldb because the number of negative samples could be set freely. I think the image_data layer is good enough for the triplet arrangement using just a txt file which could be easily generated from MATLAB, C++, python etc, what do you think?
@bhack This caffe PR hasn't dependency on my OpenCV PR, and the OpenCV has a dependency on it:) I have also attached a python script using Blender python api on normalization for any .obj files which could be further rendered by VTK toolkit or my OpenCV contrib codes. could I attach another code together with a cmakelist file which is used for rendering training data from 3D models?
@pwohlhart @mtamburrano  Thanks for your guidance on these codes, I really appreciate it!

wusongbeckham commented:
Hi, I am a beginner about the triplet network. I have a simple question about how is the label for the training and testing data?
Thanks

Wangyida commented:
@wusongbeckham  You could have a reference on the Training Data Preparation part in the readme.md in examples/triplet folder, the labels are not necessary for triplet loss because you must arrange data in advance in a sequence regarding to your task (category, pose, subcategory, etc.), and the label could also be used is you also use some softmax loss combined with triplet loss in net.

Wangyida commented:
@jackculpepper I have added your reference in header files for triplet layer, I think your locally connected layer is also very useful in caffe, how is that going on? Codes of your local layer seems work well.

bhack commented:
I only think that will be useful to have a model trained on Pascal 3d with prototxt on model zoo so will be easier for OpenCV users to train or produce descriptors for Pascal. So If you have got interesting results on this dataset you could use model zoo and leave in thos PR only Mnist as @shelhamer requested.

wusongbeckham commented:
Hi, dear Yida,
Thank for your reply and in my view for only the triplet network training,
the label is not necessary. I read the example that you set the label all
to 1.
Thanks for your help.

Song Wu
Leiden University

On Tue, Feb 23, 2016 at 3:51 AM, Yida Wang <notifications@github.com> wrote:

> @wusongbeckham <https://github.com/wusongbeckham> You could have a
> reference on the Training Data Preparation part in the readme.md in
> examples/triplet folder, the labels are not necessary for triplet loss
> because you must arrange data in advance in a sequence regarding to your
> task (category, pose, subcategory, etc.), and the label could also be used
> is you also use some softmax loss combined with triplet loss in net.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BVLC/caffe/pull/3123#issuecomment-187493163>.
>

wusongbeckham commented:
Hi, dear Yida,
I meet a strange issue that when I use gcc to compile a .cpp file, it tell
me ¨fatal error: caffe/proto/caffe.pb.h: No such file or directory ¨. I
then add caffe.pb.h by this way:

In the directory you installed Caffe to

protoc src/caffe/proto/caffe.proto --cpp_out=.
mkdir include/caffe/proto
mv src/caffe/proto/caffe.pb.h include/caffe/proto

Finally, the caffe.pb.h exists but it still have the error: fatal error:
caffe/proto/caffe.pb.h: No such file or directory

Do you know how to solve this?

The attached is my .cpp file for data_convert


On Tue, Feb 23, 2016 at 3:51 AM, Yida Wang <notifications@github.com> wrote:

> @wusongbeckham <https://github.com/wusongbeckham> You could have a
> reference on the Training Data Preparation part in the readme.md in
> examples/triplet folder, the labels are not necessary for triplet loss
> because you must arrange data in advance in a sequence regarding to your
> task (category, pose, subcategory, etc.), and the label could also be used
> is you also use some softmax loss combined with triplet loss in net.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BVLC/caffe/pull/3123#issuecomment-187493163>.
>

jackculpepper commented:
@Wangyida Glad you found it useful! Have you seen this PR:

https://github.com/BVLC/caffe/pull/3068

?

@raingo brought it up to date with a more recent master.

By the way, https://github.com/BVLC/caffe/pull/3663 is a triplet loss that has multiple bottoms, instead of interleaving anchor, positive, and negative within the batch as you do.

I also wrote a triplet loss (based on yours) but didn't create a pull request for it yet. Mine is here:

https://github.com/jackculpepper/caffe/tree/triplet_loss

I like your way of interleaving anchor, positive, and negative examples within the batch. It would be nice to have a feature that would sample K triplets given a single anchor, a few positives, and many negatives.

wusongbeckham commented:
@Wangyida
Hi, dear Yida,
I meet a strange issue that when I use gcc to compile a .cpp file, it tell
me ¨fatal error: caffe/proto/caffe.pb.h: No such file or directory ¨. I
then add caffe.pb.h by this way:

In the directory you installed Caffe to

protoc src/caffe/proto/caffe.proto --cpp_out=.
mkdir include/caffe/proto
mv src/caffe/proto/caffe.pb.h include/caffe/proto

Finally, the caffe.pb.h exists but it still have the error: fatal error:
caffe/proto/caffe.pb.h: No such file or directory

Do you know how to solve this?
The attached is my .cpp file for data_convert.

On Tue, Feb 23, 2016 at 3:51 AM, Yida Wang <notifications@github.com> wrote:

> @wusongbeckham <https://github.com/wusongbeckham> You could have a
> reference on the Training Data Preparation part in the readme.md in
> examples/triplet folder, the labels are not necessary for triplet loss
> because you must arrange data in advance in a sequence regarding to your
> task (category, pose, subcategory, etc.), and the label could also be used
> is you also use some softmax loss combined with triplet loss in net.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BVLC/caffe/pull/3123#issuecomment-187493163>.
>

pwohlhart commented:
@jackculpepper 
regarding the triplet loss: what I had in my theano implementation was for each minibatch a triplet index given to the loss layer as additional input. Each row in the index (anchor,plus,minus) specifies a triplet within the minibatch.
This way you can include a sample and calculate it's descriptor just once and then use it in more than one triplet. In my use case I had it quite often, that samples occur more than one triplet. Plus, without modifying the minibatch data, you can add more triplets and do hard negative mining within the minibatch by adding triplets to the index.

However, the main drawback (and I guess the reason why @Wangyida didn't implement it this way) is that preparing the minibatches becomes quite a bit more complex and cumbersome, because you have to keep track of what you already have in the minibatch and where it went.

Wangyida commented:
@wusongbeckham I think the best way for compiling and installing Caffe will be using cmake as described in http://caffe.berkeleyvision.org/installation.html , the caffe.pb.h file might be a generated file in regard to caffe.proto file.

wusongbeckham commented:
@Wangyida
Hi, dear yida,
I solve this problem and when I run the following to compile the code:
g++ -I/home/deep-learning/Documents/caffe-wusong/include/
-I/home/deep-learning/Documents/caffe-wusong/src/
-I/usr/local/cuda-7.5/include/ -I/usr/include/ convert_ubc_triplet_data.cpp
-o convert_ubc_triplet_data `pkg-config --cflags --libs opencv`
-L/usr/local/lib -L/usr/local/cuda/lib64 -L/usr/local/lib
-L/home/deep-learning/Documents/caffe-wusong/build/lib -lcudart -lcublas
-lcurand -lglog -lgflags -lprotobuf -lleveldb -lsnappy -llmdb
-lboost_system -lhdf5_hl -lhdf5 -lm
-L/home/deep-learning/Documents/caffe-wusong/build/lib -lcaffe
The compile is successful, but when I run, I got the following error:
./convert_ubc_triplet_data: error while loading shared libraries:
libcaffe.so.1.0.0-rc3: cannot open shared object file: No such file or
directory

do you have any idea about this error?



On Wed, Feb 24, 2016 at 3:51 PM, Yida Wang <notifications@github.com> wrote:

> @wusongbeckham <https://github.com/wusongbeckham> I think the best way
> for compiling and installing Caffe will be using cmake as described in
> http://caffe.berkeleyvision.org/installation.html , the caffe.pb.h file
> might be a generated file in regard to caffe.proto file.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BVLC/caffe/pull/3123#issuecomment-188286499>.
>

Wangyida commented:
@pwohlhart Yes, your way in Theano saves a lot of memory, I think the input image_data layer with image list seems more similar to your method loading training samples in a single batch, images will be stores in disk just once in this way, but some samples will still be duplicated in a mini batch during iterations, this is a drawback indeed. If we use DB files instead of image list style, there will be more duplicated data stored in disk after I arranged in triplet sets.

wusongbeckham commented:
@Wangyida 
Hi, dear Yida,
I want to train the triplet network to compare two local patch in the image, following the paper "Learning Local Image Descriptors with Deep Siamese and Triplet Convolutional Networks by Minimising Global Loss Functions". 
I still feel not clear that: if there is no label, how to evaluate the accuracy?
Song

Wangyida commented:
@wusongbeckham Triplet loss is used as a guidance for descriptor learning, you could join it with softmax loss for prediction or you can train a classifier based on the learnt feature from triplet loss lead training.

igorbb commented:
Sorry If I am asking trivial things here @Wangyida. But I am trying to get the overall procedure.

Can you show a example  on training on MNIST with 2 negative samples ?
I could not figure it out form the Example documentation you have. It seems it dealing with PLYs ?

igorbb commented on examples/triplet/readme.md:
```diff
@@ -0,0 +1,87 @@
+
+We will assume that you have caffe successfully compiled. If not, please refer
+to the [Installation page](../../installation.html). This example builds on the
+[MNIST tutorial](mnist.html) so it would be a good idea to read that before
```
Dead Link ?

Wangyida commented:
@igorbb I updated readme in triplet examples and the convert_triplet_data codes with instructions about how to modify data reading function and conditionals for issues not related to pose estimation like prediction in MNIST. This codes assumes that you have a basic knowledge on binary file which is the format of MNIST database, you could also have a reference on example in siamese data conversion codes which helps you figure out how binary label is utilised in data arrangement for DB files.

igorbb commented:
Thanks @Wangyida .

Reading previous comments and some of the code I think things are following into place...

So in theory one could just use the  examples/imagenet/create_imagenet.sh to convert an image list to a LMDB, and if the list is like the following ( assuming 3 negative samples), it will be okay.

```
root_path_folder/anchor_A.jpg CLASS_A
root_path_folder/positve_A.jpg CLASS_A
root_path_folder/negative_A1.jpg CLASS_NOT_A
root_path_folder/negative_A2.jpg CLASS_NOT_A
root_path_folder/negative_A3.jpg CLASS_NOT_A
...
root_path_folder/anchor_B.jpg CLASS_B
root_path_folder/positve_B.jpg CLASS_B
root_path_folder/negative_B1.jpg CLASS_NOT_B
root_path_folder/negative_B2.jpg CLASS_NOT_B
root_path_folder/negative_B3.jpg CLASS_NOT_B
....
```


And I do need to be carefurell to set batch as a multiple of 5.
The class is not really important as Triplet Losst wont be used as label, 

Are my intentions correct ?

igorbb commented:
Hi @ljw3351639 I still get Segmentation fault with TripletLossLayer<>::Forward_gpu(). Did you modified  `mutable_cpu_data()` with `mutable_gpu_data() `in src/caffe/layers/triplet_loss_layer.cu or is your file unaltered ?

Wangyida commented:
@igorbb I have tested my latest codes again on triplet loss type 0 and 1 for K40 GPU. There isn't such problem as you have described. Maybe you should pull the latest codes because I rebased all commits into a single commit.

igorbb commented:
@Wangyida I managed to get the previous commit running (with the labs Titan X) after some small modifications.
I will test the new commit Friday and let you know by the weekend.

igorbb commented:
Hey @Wangyida 
Just tested   https://github.com/Wangyida/caffe/tree/58dc67922be2052fa399ffe1a482ef43babce5fa . Now it is working fine. Maybe I had a  pre-base version.

Wangyida commented:
@igorbb Good news, there are still some CPU operation in my GPU codes. Welcome anyone boosting the efficiency of it.

wusongbeckham commented:
@Wangyida 
Dear Yida,
I defined a new loss function and when I see the source code of backward_cpu, you just use the caffe_cpu_axpby(). Could you help me to explain the details of this function?
Thanks in advance.
Song Wu

Wangyida commented:
@wusongbeckham This is a linear operation function dependent on cblas, you could just have a reference on cblas headers.

wusongbeckham commented:
Hi, dear Yida,
Do you have the code which is about how to generate the
convert_minst_triplet_data.bin. I can not find the bin file.
Thanks.
Song Wu

On Wed, Mar 9, 2016 at 1:33 AM, Yida Wang <notifications@github.com> wrote:

> @wusongbeckham <https://github.com/wusongbeckham> This is a linear
> operation function dependent on cblas, you could just have a reference on
> cblas headers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BVLC/caffe/pull/3123#issuecomment-194036998>.
>

wusongbeckham commented:
@Wangyida 
Hi, dear Yida,
Do you have the code which is about how to generate the convert_minst_triplet_data.bin. I can not find the bin file.
Thanks.
Song Wu

Wangyida commented:
The previous codes are written for early type of triplet and you could have a reference on examples/triplet/convert_triplet_data.cpp

wusongbeckham commented:
@Wangyida 
Hi, Yida,
During the training of triplet network on the MNIST in your code, the test loss is becoming bigger and bigger. Why? I think it should become smaller and smaller.
Thanks.
Song Wu

Wangyida commented:
@wusongbeckham Problem might be the fact that you are using a small batch size or the. The loss should be smaller in overall process. It also goes wrong if the samples are not properly arranged while you have selected too many negative samples. You can have a test on Losstype 0 while Losstype 1 and 2 are more efficient.

Aldream commented:
@Wangyida 
First of all, great job here! May I ask what is the status of this PR? There doesn't seem to be any activity around it now...

I just started digging into your implementation and example, but as @raingo and @igorbb suggested, it could be nice to have an end-to-end example, covering also the data preparation (a .pynb file is a good idea). Just to get started faster...

danst18 commented:
@Wangyida 
Thanks for your contribution.

I am wondering if you or other people are doing online generation of triplets and how. It seems that this is a crucial step in the triplet learning process and I can't find any way to do this in Caffe.

seanbell commented on src/caffe/proto/caffe.proto:
```diff
@@ -544,6 +545,14 @@ message ContrastiveLossParameter {
+message TripletLossParameter {
+  //margin for negative triplet
+  optional float margin = 1 [default = 1.0];
+  optional uint32 losstype = 2 [default = 1];
```
I think the code would be much more readable if this were a string or enum.

bjutdy replied:
@wusongbeckham 
I user convert_triplet_data.bin convert Minist to levedb , but it is fail ,can you give me the create_minist_triple.sh ? thank you

seanbell commented on src/caffe/proto/caffe.proto:
```diff
@@ -544,6 +545,14 @@ message ContrastiveLossParameter {
+  optional float margin = 1 [default = 1.0];
+  optional uint32 losstype = 2 [default = 1];
+  optional uint32 num_negatives = 3 [default = 3];
+  optional uint32 use_pair = 4 [default = 0];
```
This is being used like a boolean; maybe this should be made a `bool` instead?

bjutdy commented:
hi,Hi, Yida:
  can you give us the url of the train dataset?

Wangyida commented:
@bjutdy OK, I will upload my rendered training data from ShapeNet for recognition in PASCAL3D+ in dropbox, you could download it.

bjutdy commented:
@Wangyida 
thanks ! I want to test the triplet, I have minist only,but i can not user the caffe/build/examples/triple/convert_triplet_data.bin convert minist to leveldb , can you give us some suggestions ?

vtantia commented:
Hi, @Wangyida are you working on this? There should be a small change. In src/caffe/proto/caffe.proto the triplet_loss_param cannot be 143 as it has already been used for input_param. I have assigned it 146 and it is compiling fine.

vtantia commented:
Also, @Wangyida could you please share the dropbox link for the data?

aizawak commented:
Amazing work @Wangyida. I'd like to implement triplet loss on a data set of my own. How should I label my data? (I'm trying to convert the data myself and am wondering what format to use to identify positive, negative, anchor, etc.)

aizawak commented:
Is anyone else returning the following failed test:

`[----------] 2 tests from TripletLossLayerTest/0, where TypeParam = struct caffe::CPUDevice<float>
[ RUN      ] TripletLossLayerTest/0.TestForward
..\..\src\caffe\test\test_triplet_loss_layer.cpp(112): error: The difference between this->blob_top_loss_->cpu_data()[0] and loss is 1.0810106992721558, which exceeds 1e-6, where
this->blob_top_loss_->cpu_data()[0] evaluates to 0.98075640201568604,
loss evaluates to 2.0617671012878418, and
1e-6 evaluates to 9.9999999999999995e-007.
[  FAILED  ] TripletLossLayerTest/0.TestForward, where TypeParam = struct caffe::CPUDevice<float> (0 ms)
[ RUN      ] TripletLossLayerTest/0.TestGradient
C:\caffe\include\caffe/test/test_gradient_check_util.hpp(175): error: The difference between computed_gradient and estimated_gradient is 0.077058762311935425, which exceeds threshold_ * scale, where
computed_gradient evaluates to -0.43571779131889343,
estimated_gradient evaluates to -0.35865902900695801, and
threshold_ * scale evaluates to 0.0099999997764825821.
debug: (top_id, top_data_id, blob_id, feat_id)=0,0,0,35; feat = 0.52477920055389404; objective+ = 1.8039973974227905; objective- = 1.8111705780029297
C:\caffe\include\caffe/test/test_gradient_check_util.hpp(175): error: The difference between computed_gradient and estimated_gradient is 0.038115590810775757, which exceeds threshold_ * scale, where
computed_gradient evaluates to 0.17970642447471619,
estimated_gradient evaluates to 0.14159083366394043, and
threshold_ * scale evaluates to 0.0099999997764825821.
debug: (top_id, top_data_id, blob_id, feat_id)=0,0,0,37; feat = -0.57654058933258057; objective+ = 1.8086471557617187; objective- = 1.8058153390884399
[  FAILED  ] TripletLossLayerTest/0.TestGradient, where TypeParam = struct caffe::CPUDevice<float> (1 ms)
[----------] 2 tests from TripletLossLayerTest/0 (2 ms total)

[----------] 2 tests from TripletLossLayerTest/1, where TypeParam = struct caffe::CPUDevice<double>
[ RUN      ] TripletLossLayerTest/1.TestForward
..\..\src\caffe\test\test_triplet_loss_layer.cpp(112): error: The difference between this->blob_top_loss_->cpu_data()[0] and loss is 1.0810105808020773, which exceeds 1e-6, where
this->blob_top_loss_->cpu_data()[0] evaluates to 0.98075638929016384,
loss evaluates to 2.0617669700922412, and
1e-6 evaluates to 9.9999999999999995e-007.
[  FAILED  ] TripletLossLayerTest/1.TestForward, where TypeParam = struct caffe::CPUDevice<double> (1 ms)
[ RUN      ] TripletLossLayerTest/1.TestGradient
C:\caffe\include\caffe/test/test_gradient_check_util.hpp(175): error: The difference between computed_gradient and estimated_gradient is 0.077061012824835573, which exceeds threshold_ * scale, where
computed_gradient evaluates to -0.43571781438032586,
estimated_gradient evaluates to -0.35865680155549029, and
threshold_ * scale evaluates to 0.01.
debug: (top_id, top_data_id, blob_id, feat_id)=0,0,0,35; feat = 0.52477923920378089; objective+ = 1.8039973966033025; objective- = 1.8111705326344123
C:\caffe\include\caffe/test/test_gradient_check_util.hpp(175): error: The difference between computed_gradient and estimated_gradient is 0.038117944273945215, which exceeds threshold_ * scale, where
computed_gradient evaluates to 0.17970644072077691,
estimated_gradient evaluates to 0.14158849644683169, and
threshold_ * scale evaluates to 0.01.
debug: (top_id, top_data_id, blob_id, feat_id)=0,0,0,37; feat = -0.57654061773791909; objective+ = 1.808647101821613; objective- = 1.8058153318926764
[  FAILED  ] TripletLossLayerTest/1.TestGradient, where TypeParam = struct caffe::CPUDevice<double> (1 ms)
[----------] 2 tests from TripletLossLayerTest/1 (3 ms total)

[----------] 2 tests from TripletLossLayerTest/2, where TypeParam = struct caffe::GPUDevice<float>
[ RUN      ] TripletLossLayerTest/2.TestForward
..\..\src\caffe\test\test_triplet_loss_layer.cpp(112): error: The difference between this->blob_top_loss_->cpu_data()[0] and loss is 1.0810106992721558, which exceeds 1e-6, where
this->blob_top_loss_->cpu_data()[0] evaluates to 0.98075640201568604,
loss evaluates to 2.0617671012878418, and
1e-6 evaluates to 9.9999999999999995e-007.
[  FAILED  ] TripletLossLayerTest/2.TestForward, where TypeParam = struct caffe::GPUDevice<float> (14 ms)
[ RUN      ] TripletLossLayerTest/2.TestGradient
C:\caffe\include\caffe/test/test_gradient_check_util.hpp(175): error: The difference between computed_gradient and estimated_gradient is 0.077058762311935425, which exceeds threshold_ * scale, where
computed_gradient evaluates to -0.43571779131889343,
estimated_gradient evaluates to -0.35865902900695801, and
threshold_ * scale evaluates to 0.0099999997764825821.
debug: (top_id, top_data_id, blob_id, feat_id)=0,0,0,35; feat = 0.52477920055389404; objective+ = 1.8039973974227905; objective- = 1.8111705780029297
C:\caffe\include\caffe/test/test_gradient_check_util.hpp(175): error: The difference between computed_gradient and estimated_gradient is 0.038115590810775757, which exceeds threshold_ * scale, where
computed_gradient evaluates to 0.17970642447471619,
estimated_gradient evaluates to 0.14159083366394043, and
threshold_ * scale evaluates to 0.0099999997764825821.
debug: (top_id, top_data_id, blob_id, feat_id)=0,0,0,37; feat = -0.57654058933258057; objective+ = 1.8086471557617187; objective- = 1.8058153390884399
[  FAILED  ] TripletLossLayerTest/2.TestGradient, where TypeParam = struct caffe::GPUDevice<float> (593 ms)
[----------] 2 tests from TripletLossLayerTest/2 (608 ms total)

[----------] 2 tests from TripletLossLayerTest/3, where TypeParam = struct caffe::GPUDevice<double>
[ RUN      ] TripletLossLayerTest/3.TestForward
..\..\src\caffe\test\test_triplet_loss_layer.cpp(112): error: The difference between this->blob_top_loss_->cpu_data()[0] and loss is 1.0810105808020773, which exceeds 1e-6, where
this->blob_top_loss_->cpu_data()[0] evaluates to 0.98075638929016384,
loss evaluates to 2.0617669700922412, and
1e-6 evaluates to 9.9999999999999995e-007.
[  FAILED  ] TripletLossLayerTest/3.TestForward, where TypeParam = struct caffe::GPUDevice<double> (7 ms)
[ RUN      ] TripletLossLayerTest/3.TestGradient
C:\caffe\include\caffe/test/test_gradient_check_util.hpp(175): error: The difference between computed_gradient and estimated_gradient is 0.077061012824835573, which exceeds threshold_ * scale, where
computed_gradient evaluates to -0.43571781438032586,
estimated_gradient evaluates to -0.35865680155549029, and
threshold_ * scale evaluates to 0.01.
debug: (top_id, top_data_id, blob_id, feat_id)=0,0,0,35; feat = 0.52477923920378089; objective+ = 1.8039973966033025; objective- = 1.8111705326344123
C:\caffe\include\caffe/test/test_gradient_check_util.hpp(175): error: The difference between computed_gradient and estimated_gradient is 0.038117944273945215, which exceeds threshold_ * scale, where
computed_gradient evaluates to 0.17970644072077691,
estimated_gradient evaluates to 0.14158849644683169, and
threshold_ * scale evaluates to 0.01.
debug: (top_id, top_data_id, blob_id, feat_id)=0,0,0,37; feat = -0.57654061773791909; objective+ = 1.808647101821613; objective- = 1.8058153318926764
[  FAILED  ] TripletLossLayerTest/3.TestGradient, where TypeParam = struct caffe::GPUDevice<double> (592 ms)
[----------] 2 tests from TripletLossLayerTest/3 (600 ms total)`

aizawak commented on src/caffe/test/test_triplet_loss_layer:
```diff
@@ -0,0 +1,125 @@
+#include "caffe/common.hpp"
+#include "caffe/filler.hpp"
+#include "caffe/vision_layers.hpp"
+
```
1. Missing .cpp file extension
2. Is '#include "caffe/vision_layers.hpp"' needed?
3. Missing '#include "caffe/layers/triplet_loss_layer.hpp"'

antran89 commented:
cc

kwin-wang commented:
@wangyida I don't  really understand this problem. Could you describe this problem in more detail?

> @Strateus triplet loss could surely do this, but there might be a problem: a single pair wise term is joint together with several triplet sets, so your reference sample and positive samples should not differs too much. You could just remove the pair wise term in the triplet loss if the loss don't decrease as expected.
